

# Manuscripts describing BEDbase and geniml

Components of BEDbase and geniml have been published independently. Here, we showcase the tutorials organized by the manuscript that published them, so if you're reading a paper you can more easily find the relevant features. Find the citation of interest on the left to find links to tutorials published as part of each manuscript.

## How to cite BEDbase and geniml

Thanks for citing us! If you use BEDbase, geniml, or their components in your research, here are papers you can cite.

| If you use... | Please cite ... |
|---------------|-----------------|
| `region2vec` embeddings | Gharavi et al. (2021) *Bioinformatics* |
| `bedspace` search and embeddings | Gharavi et al. (2024) *Bioengineering* |
| `scEmbed` single-cell embedding framework | LeRoy et al. (2024) *NAR Genomics and Bioinformatics* |
| `geniml` region set evaluations | Zheng et al. (2024) *NAR Genomics and Bioinformatics* |
| `geniml hmm` module | Rymuza et al. (2024) *Nucleic Acis Research* |
| `bedbase` database | Unpublished |


## Full citation information for manuscripts 

<li><b>Gharavi et al. (2024). </b><i>Joint representation learning for retrieval and annotation of genomic interval sets</i> 
<br><i>Bioengineering</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.3390/bioengineering11030263">10.3390/bioengineering11030263</a></span></li>
<li><b>Zheng et al. (2024). </b><i>Methods for evaluating unsupervised vector representations of genomic regions</i> 
<br><i>Nucleic Acids Research Genomics and Bioinformatics</i>.  <span class="doi">DOI: <a href="https://doi.org/10.1093/nargab/lqae086">10.1093/nargab/lqae086</a></span></li>
<li><b>Xue et al. (2023). </b><i>Opportunities and challenges in sharing and reusing genomic interval data</i> 
<br><i>Frontiers in Genetics</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.3389/fgene.2023.1155809">10.3389/fgene.2023.1155809</a></span></li>
<li><b>Rymuza et al. (2024). </b><i>Methods for constructing and evaluating consensus genomic interval sets</i> 
<br><i>Nucleic Acids Research</i>.  <span class="doi">DOI: <a href="https://doi.org/10.1093/nar/gkae685">10.1093/nar/gkae685</a></span></li>
<li><b>LeRoy et al. (2024). </b><i>Fast clustering and cell-type annotation of scATACdata with pre-trained embeddings</i> 
<br><i>Nucleic Acids Research Genomics and Bioinformatics</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1093/nargab/lqae073">10.1093/nargab/lqae073</a></span></li>
<li><b>Gu et al. (2021). </b><i>Bedshift: perturbation of genomic interval sets</i> 
<br><i>Genome Biology</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1186/s13059-021-02440-w">10.1186/s13059-021-02440-w</a></span></li>
<li><b>Gharavi et al. (2021). </b><i>Embeddings of genomic region sets capture rich biological associations in low dimensions</i> 
<br><i>Bioinformatics</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1093/bioinformatics/btab439">10.1093/bioinformatics/btab439</a></span></li>


