

# Manuscripts describing BEDbase and geniml

Components of BEDbase and geniml have been published independently. Here, we showcase the tutorials organized by the manuscript that published them, so if you're reading a paper you can more easily find the relevant features. Find the citation of interest on the left to find links to tutorials published as part of each manuscript.

## How to cite BEDbase and geniml

Thanks for citing us! If you use BEDbase, geniml, or their components in your research, here are papers you can cite.

| If you use...                             | Please cite ... |
|-------------------------------------------|-----------------|
| `region2vec` embeddings                   | Gharavi et al. (2021) *Bioinformatics* |
| `bedspace` search and embeddings          | Gharavi et al. (2024) *Bioengineering* |
| `scEmbed` single-cell embedding framework | LeRoy et al. (2024) *NAR Genomics and Bioinformatics* |
| `geniml` region set evaluations           | Zheng et al. (2024) *NAR Genomics and Bioinformatics* |
| `geniml hmm` module                       | Rymuza et al. (2024) *Nucleic Acis Research* |
| `bedbase` database                        | Unpublished |
| `tokenizers` module                       | LeRoy et al. (2025) *arXiv* |
| `atacformer` foundation model             | LeRoy et al. (2025) *bioRxiv* |


## Full citation information for manuscripts

- **LeRoy et al. (2025).** *Atacformer: A transformer-based foundation model for analysis and interpretation of ATAC-seq data*.
  *bioRxiv*. [10.1101/2025.11.03.685753](https://doi.org/10.1101/2025.11.03.685753)

- **LeRoy et al. (2025).** *Fast, memory-efficient genomic interval tokenizers for modern machine learning*.
  *arXiv*. [10.48550/arXiv.2511.01555](https://doi.org/10.48550/arXiv.2511.01555) 

- **Gharavi et al. (2024).** *Joint representation learning for retrieval and annotation of genomic interval sets*.
  *Bioengineering*. [10.3390/bioengineering11030263](http://dx.doi.org/10.3390/bioengineering11030263)

- **Zheng et al. (2024).** *Methods for evaluating unsupervised vector representations of genomic regions*.
  *Nucleic Acids Research Genomics and Bioinformatics*. [10.1093/nargab/lqae086](https://doi.org/10.1093/nargab/lqae086)

- **Xue et al. (2023).** *Opportunities and challenges in sharing and reusing genomic interval data*.
  *Frontiers in Genetics*. [10.3389/fgene.2023.1155809](http://dx.doi.org/10.3389/fgene.2023.1155809)

- **Rymuza et al. (2024).** *Methods for constructing and evaluating consensus genomic interval sets*.
  *Nucleic Acids Research*. [10.1093/nar/gkae685](https://doi.org/10.1093/nar/gkae685)

- **LeRoy et al. (2024).** *Fast clustering and cell-type annotation of scATAC data with pre-trained embeddings*.
  *Nucleic Acids Research Genomics and Bioinformatics*. [10.1093/nargab/lqae073](http://dx.doi.org/10.1093/nargab/lqae073)

- **Gu et al. (2021).** *Bedshift: perturbation of genomic interval sets*.
  *Genome Biology*. [10.1186/s13059-021-02440-w](http://dx.doi.org/10.1186/s13059-021-02440-w)

- **Gharavi et al. (2021).** *Embeddings of genomic region sets capture rich biological associations in low dimensions*.
  *Bioinformatics*. [10.1093/bioinformatics/btab439](http://dx.doi.org/10.1093/bioinformatics/btab439)


